activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,184 @@ def valid_message?(signed_message)
+    #
+    # * <tt>:expires_at</tt> - Explicit expiring time for the signed message.
+    # * <tt>:expires_in</tt> - Relative time the message should
+    #   expire after (e.g. 1.hour).
+    # * <tt>:for</tt> - The signed message purpose confines usage to places with
+    #   the same purpose. Defaults to a universally usable message.
+    #
+    # ==== Examples
+    #
+    # Messages can be signed to ensure that the data hasn't been tampered with.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   verifier.generate 'a private message' # => "BAhJIhRwcml2YXRlLW1lc3NhZ2UGOgZFVA==--e2d724331ebdee96a10fb99b089508d1c72bd772"
-    def generate(value)
-      data = encode(@serializer.dump(value))
-      "#{data}--#{generate_digest(data)}"
+    #   verifier.generate('a private message')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIOdW5pdmVyc2FsBjsGVDoIZXhwSSIdMjAxNS0wOC0wOVQxODowNjowMy44NzhaBjsGVA==.ZmRlYjhlMTkxNGI0NGQ4OWRkNDdjZWM5YTY1NmQ3MTM5OWY5NzRmMQ=="
+    #
+    # Messages can be given a purpose to bump the security up some more.
+    # This way evildoers can't reuse a message generated for the sign-up form
+    # on any other page:
+    #
+    #   signup_message = verifier.generate('a private message', for: 'signup_form')
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIQc2lnbnVwX2Zvcm0GOwZUOghleHBJIh0yMDE1LTA4LTA5VDE4OjA1OjEwLjM2N1oGOwZU.YTMxMzIzMDc3YTI3ZDkwM2FmNWQ1MDgwNzAxNTgwMjNjNzI2NDE0Zg=="
+    #
+    #   verifier.verify(signup_message, for: 'signup_form')
+    #   # => 'a private message'
+    #
+    #   verifier.verified(signup_message, for: 'something_else') # => nil
+    #
+    # A message can also be given an expiration time with either +expires_in+
+    # or +expires_at+. Useful when a message should not be accessible indefinitely.
+    # The +expires_in+ option accepts a relative time, which is resolved to an
+    # explicit expiry time, when the message is signed:
+    #
+    #   expiring_message = verifier.generate('a private message', for: 'remember_me', expires_in: 1.hour)
+    #   # => "BAh7B0kiCHR5cAY6BkVUSSIISldUBjsAVEkiCGFsZwY7AFRJIglTSEExBjsAVA==.BAh7CDoIcGxkSSIWYSBwcml2YXRlIG1lc3NhZ2UGOgZFVDoIZm9ySSIQcmVtZW1iZXJfbWUGOwZUOghleHBJIh0yMDE1LTA3LTA5VDE5OjAzOjEwLjQ3MVoGOwZU.MmQxNzkxMmY2MGU5ZmQ2YTlhNmRhYmYyNmEzNDUzNjE0ZDdjNzg2Yw=="
+    #
+    #   # Within 1 hour...
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # => "a private message"
+    #
+    #   # After 1 hour...
+    #   verifier.verify(expiring_message, for: 'remember_me')
+    #   # raises ActiveSupport::Claims::ExpiredClaims
+    #
+    # A message can also be set to expire at an explicit time with +expires_at+:
+    #
+    #   expiring_message = verifier.generate('a private message', for: 'something', expires_at: Date.tomorrow.midnight)
```

Comment:
I agree, let's keep it to this one example of creating a message using `expires_at` - readers already know what it looks like when a message expires, so we don't have to show that again.